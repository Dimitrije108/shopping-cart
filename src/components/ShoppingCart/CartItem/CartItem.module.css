.horizontalLine {
	margin: 1rem 0;
  border: 0;
  height: 1px;
  background-color: rgb(104, 104, 104);
}

.cartItem {
	display: grid;
	grid-template-columns: 1fr 1.5fr 0.2fr 1fr;
	align-items: end;
}

.imgCont{
	height: 125px;
	width: 250px;

	cursor: pointer;
}

.imgCont > img{
	height: 100%;
	width: 100%;
	object-fit: cover;
}

.name {
	margin: 0;
	padding: 4px 8px 0 8px;

	font-family: DINNext;
	color: rgb(221, 221, 221);
	font-size: 1.3rem;

  transition: all 0.5s ease-in-out;
}

.name:hover {
  text-decoration: underline rgb(221, 221, 221);
	text-decoration-thickness: 1px;
	cursor: pointer;
}

.imperial {
	margin: 0;
	padding: 0 9px;

  color: rgb(179, 179, 179);
  font-family: Imperial;
  font-size: 0.6rem;
  letter-spacing: 1.5px;

	cursor: pointer;
}

.naboo {
	margin: 0;
	padding: 0 9px 4px 9px;

  color: rgb(179, 179, 179);
  font-family: Naboo;
  font-size: 0.4rem;
  letter-spacing: 1.5px;

	cursor: pointer;
}

.priceCont {
	height: 100%;

	display: flex;
	flex-direction: column;
	justify-content: space-between;
	align-items: flex-end;
}

.delBtn {
	margin: 0;
	padding: 0;
	border: none;
}

.delBtn:hover {
	margin: 0;
	padding: 0;
	border: none;
}

.delBtn > img {
	vertical-align: middle;
}

.priceCont > p {
	font-size: 1.3rem;
	text-align: end;
}

.itemQuantity {
	font-size: 0.9rem;
}