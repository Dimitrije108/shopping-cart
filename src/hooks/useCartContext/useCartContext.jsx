import { useState, useEffect, createContext, useContext } from 'react';

const CartContext = createContext({
  cart: null,
	limitExceededMsg: "", 
  addToCart: () => {},
	removeFromCart: () => {},
});
// Shopping cart functionality passed on to all components 
// via context provider
export function useShoppingCart() {
	// Retrieve cart from storage if available
	const storedCart = JSON.parse(localStorage.getItem('cart'));
	const initCart = storedCart ? storedCart : [];

	const [cart, setCart] = useState(initCart);

	useEffect(() => {
		localStorage.setItem('cart', JSON.stringify(cart));
	}, [cart]);
	// Warning message for exceeding the ship quantity limit
	const limitExceededMsg = 
		"For orders larger than 20 ships per model please contact us directly!";
	// Add an item or adjust it's quantity
	const addToCart = (item) => {
		// check if the provided item is already inside the cart
		const isDuplicate = cart.find((ship) => ship.id === item.id);
		// handle duplicate items
		if (isDuplicate) {
			const totalQuantity = isDuplicate.quantity + item.quantity;
			// handle exceeding item quantity
			if (totalQuantity > 20) {
				return false;
			};
			// adjust quantity accordingly
			const addQuantity = cart.map((ship) => {
				if (ship.id === item.id) {
					return { ...ship, quantity: ship.quantity + item.quantity };
				}
				return ship;
			});
			setCart(addQuantity);
			return true;
		} else {
			// add unique item to the cart
			setCart([...cart, item]);
			return true;
		};
	};
	// Remove item instance from the cart
	const removeFromCart = (id) => {
		const removed = cart.filter((ship) => ship.id !== id);
		setCart(removed);
	};

	return { 
		cart, 
		limitExceededMsg, 
		addToCart, 
		removeFromCart,
	};
};

export const useCartContext = () => useContext(CartContext);

export default CartContext;