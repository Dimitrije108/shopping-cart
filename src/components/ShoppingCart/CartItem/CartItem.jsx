import { Link } from "react-router-dom";
import PropTypes from 'prop-types';
import { useCartContext } from '../../../hooks/useCartContext/useCartContext';
import useItemQuantity from '../../../hooks/useItemQuantity/useItemQuantity';
import ItemQuantity from '../../ItemQuantity/ItemQuantity';
import formatNumber from '../../../formatNumber/formatNumber';
import styles from './CartItem.module.css';

export default function CartItem({ 
	id, 
	img, 
	name = "Default", 
	category = "default", 
	price = "100000", 
	shipQuantity = 1,
}) {
	const { 
		removeFromCart, 
		adjustItemQuantity, 
	} = useCartContext();

	const {
		quantity,
		increaseQuantity, 
		decreaseQuantity, 
		changeQuantity, 
		resetQuantity,
	} = useItemQuantity(shipQuantity);

	const handleQuantityChange = (newQuan) => {
		adjustItemQuantity(id, newQuan);
	};

	const linkTo = `/starships/${category}/${id}`;

	return (
		<div className={styles.cartItemCont} data-testid="cartItem">
			<hr className={styles.horizontalLine} />
			<div className={styles.cartItem}>
				<Link to={linkTo}>
					<div className={styles.imgCont}>
						<img src={img} alt={`${name} starship`} />
					</div>
				</Link>
				<Link to={linkTo}>
					<div className={styles.nameCont}>
						<h2 className={styles.name}>{name}</h2> 
						<div>
							<p className={styles.imperial}>{name}</p>
							<p className={styles.naboo}>{name}</p>
						</div>
					</div>
				</Link>
				<ItemQuantity
					quantity={quantity}
					increase={() => increaseQuantity(handleQuantityChange)}
					decrease={() => decreaseQuantity(handleQuantityChange)}
					change={(e) => changeQuantity(e, handleQuantityChange)}
					reset={() => resetQuantity(handleQuantityChange)}
				/>
				<div className={styles.priceCont}>
					<button 
						className={styles.delBtn}
						onClick={() => removeFromCart(id)}
					>
						<img src="/src/assets/trash.svg" alt="trashcan remove icon" width={28} height={28} />
					</button>
					<p>
						{price === "Price on Request" 
						? price
						: <>
								<span className={styles.itemQuantity}>
									{`${quantity}x `}
								</span>
								{formatNumber(quantity * price)}
							</>
						}
					</p>
				</div>
			</div>
		</div>
	)
};

CartItem.propTypes = {
	id: PropTypes.string.isRequired,
	img: PropTypes.string.isRequired,
	name: PropTypes.string,
	category: PropTypes.string,
	price: PropTypes.string,
	shipQuantity: PropTypes.number,
};
