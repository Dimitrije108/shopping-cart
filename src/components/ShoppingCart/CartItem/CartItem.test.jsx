import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import CartItem from "./CartItem";
import { useCartContext } from "../../../hooks/useCartContext/useCartContext";

vi.mock("../../../hooks/useCartContext/useCartContext", () => ({
  useCartContext: vi.fn(() => ({
		removeFromCart: vi.fn(),
		adjustItemQuantity: vi.fn(),
	})),
}));

const mockCartItem = {
	id: "6429291f021f17e13fbc1db9",
	img: "https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg",
	name: "Millennium Falcon",
	category: "starfighter",
	price: "100000",
	quantity: 1,
};

const mockCartItem2 = {
	id: "6429291f021f17e13fbc1e3c",
	img: "https://lumiere-a.akamaihd.net/v1/images/X-Wing-Fighter_47c7c342.jpeg",
	name: "X-wing Starfighter",
	category: "starfighter",
	price: "149999",
	quantity: 3,
};

describe("Cart Item component", () => {
	it("ship card component snapshot", () => {
		const cartItem = render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem.id}
					img={mockCartItem.img}
					name={mockCartItem.name}
					category={mockCartItem.category}
					price={mockCartItem.price}
					shipQuantity={mockCartItem.quantity}
				/>
			</MemoryRouter>
		)

		expect(cartItem).toMatchSnapshot();
	});

	it("renders cart item component correctly", () => {
		render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem.id}
					img={mockCartItem.img}
					name={mockCartItem.name}
					category={mockCartItem.category}
					price={mockCartItem.price}
					shipQuantity={mockCartItem.quantity}
				/>
			</MemoryRouter>
		)

		const itemLink1 = screen.getByRole("link", { name: "Millennium Falcon starship" });
		const shipImg = screen.getByRole("img", { name: "Millennium Falcon starship" });
		const heading = screen.getByRole("heading", { name: "Millennium Falcon" });
		const itemLink2 = heading.closest("a");
		const quantityInput = screen.getByRole('spinbutton');
		const removeBtn = screen.getByRole('button', { name: "trashcan remove icon" });
		const removeBtnIcon = screen.getByRole('img', { name: "trashcan remove icon" });

		expect(itemLink1).toBeInTheDocument();
		expect(shipImg).toHaveAttribute("src", "https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg");
		expect(heading).toBeInTheDocument();
		expect(itemLink2).toBeInTheDocument();
		expect(quantityInput).toBeInTheDocument();
		expect(removeBtn).toBeInTheDocument();
		expect(removeBtnIcon).toHaveAttribute("src", "/src/assets/trash.svg");
	});

	it("render 'Price on Request' when price is unavailable", () => {
		render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem.id}
					img={mockCartItem.img}
					name={mockCartItem.name}
					category={mockCartItem.category}
					price="Price on Request"
					shipQuantity={mockCartItem.quantity}
				/>
			</MemoryRouter>
		)

		const price = screen.getByText('Price on Request');

		expect(price).toHaveTextContent('Price on Request');
	});

	it("render price when price is available", () => {
		render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem.id}
					img={mockCartItem.img}
					name={mockCartItem.name}
					category={mockCartItem.category}
					price={mockCartItem.price}
					shipQuantity={mockCartItem.quantity}
				/>
			</MemoryRouter>
		)

		const price = screen.getByText("100,000");

		expect(price).toHaveTextContent("100,000");
	});

	it("render correct price based on quantity", () => {
		render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem2.id}
					img={mockCartItem2.img}
					name={mockCartItem2.name}
					category={mockCartItem2.category}
					price={mockCartItem2.price}
					shipQuantity={mockCartItem2.quantity}
				/>
			</MemoryRouter>
		)

		const price = screen.getByText("449,997");

		expect(price).toHaveTextContent("449,997");
	});

	it("remove button calls removeFromCart functionality", async () => {
		const user = userEvent.setup();

		const mockRemoveFromCart = vi.fn();

		useCartContext.mockReturnValue({
			removeFromCart: mockRemoveFromCart,
		});

		render(
			<MemoryRouter>
				<CartItem 
					id={mockCartItem2.id}
					img={mockCartItem2.img}
					name={mockCartItem2.name}
					category={mockCartItem2.category}
					price={mockCartItem2.price}
					shipQuantity={mockCartItem2.quantity}
				/>
			</MemoryRouter>
		)

		const removeBtn = screen.getByRole("button", { name: "trashcan remove icon" });
		await user.click(removeBtn);

		expect(mockRemoveFromCart).toHaveBeenCalledOnce();
		expect(mockRemoveFromCart).toHaveBeenCalledWith(mockCartItem2.id);
	});
})