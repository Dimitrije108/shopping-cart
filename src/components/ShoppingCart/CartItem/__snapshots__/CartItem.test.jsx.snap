// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Item component > cart item component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_cartItemCont_6dc023"
        data-testid="cartItem"
      >
        <hr
          class="_horizontalLine_6dc023"
        />
        <div
          class="_cartItem_6dc023"
        >
          <a
            data-discover="true"
            href="/starships/starfighter/6429291f021f17e13fbc1db9"
          >
            <div
              class="_imgCont_6dc023"
            >
              <img
                alt="Millennium Falcon starship"
                src="https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg"
              />
            </div>
          </a>
          <a
            data-discover="true"
            href="/starships/starfighter/6429291f021f17e13fbc1db9"
          >
            <div
              class="_nameCont_6dc023"
            >
              <h2
                class="_name_6dc023"
              >
                Millennium Falcon
              </h2>
              <div>
                <p
                  class="_imperial_6dc023"
                >
                  Millennium Falcon
                </p>
                <p
                  class="_naboo_6dc023"
                >
                  Millennium Falcon
                </p>
              </div>
            </div>
          </a>
          <div
            class="_cont_3c2471"
          >
            <button
              class="_leftBtn_3c2471"
            >
              -
            </button>
            <input
              class="_quantityInput_3c2471"
              max="20"
              min="1"
              type="number"
              value="1"
            />
            <button
              class="_rightBtn_3c2471"
            >
              +
            </button>
          </div>
          <div
            class="_priceCont_6dc023"
          >
            <button
              class="_delBtn_6dc023"
            >
              <img
                alt="trashcan remove icon"
                height="28"
                src="/src/assets/trash.svg"
                width="28"
              />
            </button>
            <p>
              <span
                class="_itemQuantity_6dc023"
              >
                1x 
              </span>
              100,000
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_cartItemCont_6dc023"
      data-testid="cartItem"
    >
      <hr
        class="_horizontalLine_6dc023"
      />
      <div
        class="_cartItem_6dc023"
      >
        <a
          data-discover="true"
          href="/starships/starfighter/6429291f021f17e13fbc1db9"
        >
          <div
            class="_imgCont_6dc023"
          >
            <img
              alt="Millennium Falcon starship"
              src="https://lumiere-a.akamaihd.net/v1/images/millennium-falcon-main-tlj-a_7cf89d3a.jpeg"
            />
          </div>
        </a>
        <a
          data-discover="true"
          href="/starships/starfighter/6429291f021f17e13fbc1db9"
        >
          <div
            class="_nameCont_6dc023"
          >
            <h2
              class="_name_6dc023"
            >
              Millennium Falcon
            </h2>
            <div>
              <p
                class="_imperial_6dc023"
              >
                Millennium Falcon
              </p>
              <p
                class="_naboo_6dc023"
              >
                Millennium Falcon
              </p>
            </div>
          </div>
        </a>
        <div
          class="_cont_3c2471"
        >
          <button
            class="_leftBtn_3c2471"
          >
            -
          </button>
          <input
            class="_quantityInput_3c2471"
            max="20"
            min="1"
            type="number"
            value="1"
          />
          <button
            class="_rightBtn_3c2471"
          >
            +
          </button>
        </div>
        <div
          class="_priceCont_6dc023"
        >
          <button
            class="_delBtn_6dc023"
          >
            <img
              alt="trashcan remove icon"
              height="28"
              src="/src/assets/trash.svg"
              width="28"
            />
          </button>
          <p>
            <span
              class="_itemQuantity_6dc023"
            >
              1x 
            </span>
            100,000
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
